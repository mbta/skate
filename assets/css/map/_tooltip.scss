.c-map-tooltip::before {
  border-top-color: black;
}

.c-map-tooltip {
  background-color: black;
  border-color: black;
  color: white;
  opacity: 100% !important;
  font-size: 14px;
  padding: 4px 8px 4px 8px;
  font-family: var(--bs-body-font-family);
  border-radius: var(--bs-border-radius);
}
