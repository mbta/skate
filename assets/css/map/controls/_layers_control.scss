.c-layers-control {
  position: relative;
}

.c-layers-control__button {
  @include button-icon(20px);
  background-color: $white;
  flex: 0;
  padding: 5px;
  span svg {
    stroke: none;
  }

  &--selected {
    background-color: $color-gray-200;
  }
  &:focus-visible {
    outline: revert;
  }
}

.c-layers-control__pill {
  position: absolute;
  top: -6px;
  left: -6px;
  width: 22px;
  height: 13px;
  border-radius: 16px;

  background-color: $color-strawberry-500;

  text-align: center;
  color: $white;
  font-size: 11px;
  line-height: 1;
  font-weight: var(--font-weight-semi);
  user-select: none;
}

.c-layers-control__content {
  position: absolute;
  top: 0;
  right: calc(100% + 10px);
  width: max-content;
  border-radius: 8px;

  background-color: $white;

  font-family: $font-family;
  font-size: var(--font-size-m);

  .list-group {
    margin: 0;

    .list-group-item {
      padding-top: 0;
      padding-bottom: 0;

      div {
        padding-top: 0.5rem;
        padding-bottom: 0.5rem;

        h2 {
          margin: 0;
          font-size: var(--font-size-m);
          font-weight: var(--font-weight-bold);
        }
      }
    }
  }
}
