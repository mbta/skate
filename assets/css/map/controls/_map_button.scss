@use "../../color/tokens_2024" as tokens;

.c-map-button {
  box-shadow: 0 0 0 2px rgb($black, 0.15);
  --bs-btn-font-size: 0.625rem;
  --bs-btn-border-radius: 0.2rem;
  --bs-btn-color: #{$black};
  --bs-btn-bg: #{$white};
  --bs-btn-hover-color: #{$black};
  --bs-btn-hover-bg: #{tokens.$eggplant-50};
  --bs-btn-active-color: #{tokens.$eggplant-500};
  --bs-btn-active-bg: #{$white};
  --bs-btn-focus-box-shadow: 0 0 0 2px #{rgb(tokens.$eggplant-500, 0.5)};
  --bs-btn-disabled-color: #{tokens.$gray-700};
  --bs-btn-disabled-bg: #{tokens.$gray-100};
  --bs-btn-disabled-opacity: 1;

  width: 2.375rem;
  height: 2.375rem;
}

.c-map-button--sm {
  width: 2rem;
  height: 2rem;
}

.c-map-button--lg {
  width: 3rem;
  height: 3rem;
}
