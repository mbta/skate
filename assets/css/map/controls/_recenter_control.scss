.c-recenter-control {
  path {
    fill: none;
    stroke: $color-button-secondary;
    stroke-width: 2;
  }

  &[data-is-active="true"] {
    background-color: $color-button-disabled;
    color: $color-font-grey;
    cursor: default;

    path {
      stroke: $color-font-grey;
    }
  }
}
