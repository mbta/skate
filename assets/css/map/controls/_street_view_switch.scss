// BEM-PREFIX: LC = Leaflet Control
.lc-street-view-switch {
  // #region Position
  position: absolute;
  // Leaflet uses px for positioning of their controls so we use px here to match
  top: 10px;
  right: 55px;
  // #endregion

  // #region Look
  user-select: none;

  color: $color-eggplant-700;
  background-color: $white;
  // #endregion

  // #region Layout
  display: flex;
  align-items: center;
  gap: 4px;

  font-size: 1rem;
  line-height: 1;

  padding: 8px;

  border: 1px solid $color-gray-400;
  border-radius: 7px;
  // #endregion

  // #region Sub-Components
  &__label-icon svg {
    fill: currentColor;

    height: 1em;
    height: calc(1em - 2px);
    min-width: 16px;
  }

  &__input {
    // Shrinkwrap the wrapper component to the input component so that flexbox
    // can align this element properly
    min-height: unset;
    height: min-content;

    // Remove layout offsets
    padding-left: 0;
    margin-bottom: 0;

    // Remove Negative Margins so shrinkwrap works
    .form-check-input {
      margin-inline: 4px;
    }
  }
  // #endregion

  // #region Media Queries
  @media screen and (max-width: 800px) {
    & {
      // Change size of icon and switch when on a smaller screen
      font-size: var(--font-size-s);
      padding: 4px;
    }

    &__label-text {
      @include visually-hidden;
    }

    &__input .form-check-input {
      // Move street view icon closer to switch
      margin-inline-start: 0;
    }
  }
  // #endregion
}
