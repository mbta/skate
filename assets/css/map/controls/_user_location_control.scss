.c-user-location-button {
  @extend .inherit-box, .border-box;

  color: $color-button-secondary;

  &[disabled] {
    color: $color-font-grey;
    background-color: $color-button-disabled;
  }

  &,
  &__svg {
    width: 100%;
    height: 100%;

    display: flex;
    align-items: center;
    justify-content: center;
  }

  &__svg svg {
    width: 20px;
    height: 20px;
    fill: currentColor;
  }
}

// Extend all styles in the context of a leaflet-bar to override leaflet CSS
// until we can use CSS layers for specificity
.leaflet-bar .c-user-location-button {
  @extend .c-user-location-button;
}
