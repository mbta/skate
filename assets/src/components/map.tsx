import Leaflet, { Map as LeafletMap, Marker } from "leaflet"
import "leaflet-defaulticon-compatibility" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import React, {
  MutableRefObject,
  ReactElement,
  useEffect,
  useRef,
  useState,
} from "react"
import vehicleAdherenceDisplayClass from "../helpers/vehicleAdherenceDisplayClass"
import { status } from "../models/vehicleStatus"
import { Vehicle } from "../realtime.d"

interface Props {
  vehicle: Vehicle
}

interface State {
  map: LeafletMap | null
  vehicleIcon: Marker | null
  vehicleLabel: Marker | null
}

const iconAnchor: [number, number] = [12, 12]

export const updateMap = (
  { vehicle }: Props,
  {
    map,
    vehicleIcon,
    vehicleLabel,
  }: {
    map: LeafletMap
    vehicleIcon: Marker
    vehicleLabel: Marker
  }
): void => {
  const { bearing, headwaySpacing, label: text, latitude, longitude } = vehicle
  const zoom = map.getZoom()

  const icon = Leaflet.divIcon({
    className: `m-vehicle-map__icon ${vehicleAdherenceDisplayClass(
      headwaySpacing,
      status(vehicle)
    )}`,
    html: `<svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="m-vehicle-properties-panel__schedule-adherence-status-icon"
          d="m10 2.7-6.21 16.94a2.33 2.33 0 0 0 1.38 3 2.36 2.36 0 0 0 1.93-.14l4.9-2.67 4.89 2.71a2.34 2.34 0 0 0 3.34-2.8l-5.81-17a2.34 2.34 0 0 0 -4.4 0z"
          transform-origin="${iconAnchor[0]}px ${iconAnchor[1]}px"
          transform="rotate(${bearing})"
        />
      </svg>`,
    iconAnchor,
  })

  const label = Leaflet.divIcon({
    className: "m-vehicle-map__label",
    html: `<svg>
            <text class="m-vehicle-icon__label">${text}</text>
          </svg>`,
    iconAnchor: [12, -24],
  })

  map.setView([latitude, longitude], zoom)

  vehicleIcon.setLatLng([latitude, longitude])
  vehicleLabel.setLatLng([latitude, longitude])

  vehicleIcon.setIcon(icon)
  vehicleLabel.setIcon(label)
}

const Map = (props: Props): ReactElement<HTMLDivElement> => {
  const containerRef: MutableRefObject<HTMLDivElement | null> = useRef(null)
  const [state, updateState] = useState<State>({
    map: null,
    vehicleIcon: null,
    vehicleLabel: null,
  })

  useEffect(() => {
    if (!containerRef.current) {
      return
    }

    const map =
      state.map ||
      Leaflet.map(containerRef.current, {
        layers: [
          Leaflet.tileLayer(
            `https://mbta-map-tiles-dev.s3.amazonaws.com/osm_tiles/{z}/{x}/{y}.png`,
            {
              attribution:
                '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
            }
          ),
        ],
        zoom: 16,
      })

    const vehicleIcon =
      state.vehicleIcon ||
      Leaflet.marker([props.vehicle.latitude, props.vehicle.longitude]).addTo(
        map
      )

    const vehicleLabel =
      state.vehicleLabel ||
      Leaflet.marker([props.vehicle.latitude, props.vehicle.longitude]).addTo(
        map
      )

    updateMap(props, { map, vehicleIcon, vehicleLabel })

    updateState({ map, vehicleIcon, vehicleLabel })
  }, [props, containerRef])

  return (
    <div
      id="id-vehicle-map"
      ref={container => (containerRef.current = container)}
      className="m-vehicle-map"
    />
  )
}

export default Map
