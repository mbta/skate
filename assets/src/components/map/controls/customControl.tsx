import Leaflet, { ControlOptions, ControlPosition } from "leaflet"
import { ReactNode, useEffect, useState } from "react"
import ReactDOM from "react-dom"
import { useMap } from "react-leaflet"

const positionQuerySelector = (position: ControlPosition) => {
  switch (position) {
    case "topleft":
      return ".leaflet-top.leaflet-left"
    case "topright":
      return ".leaflet-top.leaflet-right"
    case "bottomleft":
      return ".leaflet-bottom.leaflet-left"
    case "bottomright":
      return ".leaflet-bottom.leaflet-right"
  }
}

export const CustomControl = ({
  position,
  children,
  className,
  insertAfterSelector,
}: ControlOptions & {
  children: ReactNode
  className: string
  insertAfterSelector?: string
}): JSX.Element | null => {
  const map = useMap()
  const portalParent = map
    .getContainer()
    .querySelector(
      `.leaflet-control-container ${
        position ? positionQuerySelector(position) : ""
      }`
    )
  const [portalElement, setPortalElement] = useState<HTMLElement | null>(null)

  useEffect(() => {
    if (!portalElement) {
      setPortalElement(document.createElement("div"))
    }

    if (portalParent && portalElement) {
      portalElement.className = `leaflet-control ${className}`

      const elementToInsertAfter = insertAfterSelector
        ? portalParent.querySelector(insertAfterSelector)
        : null
      if (elementToInsertAfter) {
        elementToInsertAfter.after(portalElement)
      } else {
        portalParent.append(portalElement)
      }
      Leaflet.DomEvent.disableClickPropagation(portalElement)
    }

    return () => portalElement?.remove()
  }, [portalElement, portalParent, className, insertAfterSelector])

  return portalElement ? ReactDOM.createPortal(children, portalElement) : null
}
