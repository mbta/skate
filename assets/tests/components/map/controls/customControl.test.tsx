import { describe, test, expect } from "@jest/globals"
import { render } from "@testing-library/react"
import "@testing-library/jest-dom/jest-globals"
import React, { ReactNode } from "react"
import { MapContainer, ZoomControl } from "react-leaflet"
import { CustomControl } from "../../../../src/components/map/controls/customControl"
import { FullscreenControl } from "../../../../src/components/map"

const mapWrapper = ({ children }: { children: ReactNode }) => (
  <MapContainer
    center={[0, 0]}
    zoom={0}
    attributionControl={false}
    zoomControl={false}
  >
    {children}
  </MapContainer>
)

const customControlContents = () => {
  return <button className="test-control-child-1">Click Me</button>
}

describe("CustomControls", () => {
  test("when a position is specified, puts the control in the expected corner", () => {
    const { container } = render(
      <>
        <CustomControl className="test-control" position="topright">
          {customControlContents()}
        </CustomControl>
      </>,
      {
        wrapper: mapWrapper,
      }
    )

    expect(
      container.querySelector(".leaflet-control-container")
    ).toMatchSnapshot()
  })

  test("when insertAfterSelector is specified and there is a node matching that selector, then the control is added after it", () => {
    const { container } = render(
      <>
        <ZoomControl position="topright" />

        <FullscreenControl position="topright" />
        <CustomControl
          className="test-control"
          position="topright"
          insertAfterSelector=".leaflet-control-zoom"
        >
          {customControlContents()}
        </CustomControl>
      </>,
      {
        wrapper: mapWrapper,
      }
    )

    expect(
      container.querySelector(".leaflet-control-container")
    ).toMatchSnapshot()
  })

  test("when insertAfterSelector is specified and there is not a node matching that selector, then the control is added at the end of the leaflet-container", () => {
    const { container } = render(
      <>
        <ZoomControl position="topright" />

        <FullscreenControl position="topright" />
        <CustomControl
          className="test-control"
          position="topright"
          insertAfterSelector=".non-existent-selector"
        >
          {customControlContents()}
        </CustomControl>
      </>,
      {
        wrapper: mapWrapper,
      }
    )

    expect(
      container.querySelector(".leaflet-control-container")
    ).toMatchSnapshot()
  })

  test("when insertFirst is specified, then the control is added at the beginning of the leaflet-container", () => {
    const { container } = render(
      <>
        <ZoomControl position="topright" />

        <FullscreenControl position="topright" />
        <CustomControl className="test-control" position="topright" insertFirst>
          {customControlContents()}
        </CustomControl>
      </>,
      {
        wrapper: mapWrapper,
      }
    )

    expect(
      container.querySelector(
        ".leaflet-control-container .test-control:first-child"
      )
    ).toBeInTheDocument()
  })
})
