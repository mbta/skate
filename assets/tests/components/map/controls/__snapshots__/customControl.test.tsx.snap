// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomControls when a position is specified, puts the control in the expected corner 1`] = `
<div
  class="leaflet-control-container"
>
  <div
    class="leaflet-top leaflet-left"
  />
  <div
    class="leaflet-top leaflet-right"
  >
    <div
      class="leaflet-control test-control"
    >
      <button
        class="test-control-child-1"
      >
        Click Me
      </button>
    </div>
  </div>
  <div
    class="leaflet-bottom leaflet-left"
  />
  <div
    class="leaflet-bottom leaflet-right"
  />
</div>
`;

exports[`CustomControls when insertAfterSelector is specified and there is a node matching that selector, then the control is added after it 1`] = `
<div
  class="leaflet-control-container"
>
  <div
    class="leaflet-top leaflet-left"
  />
  <div
    class="leaflet-top leaflet-right"
  >
    <div
      class="leaflet-control-zoom leaflet-bar leaflet-control"
    >
      <a
        aria-disabled="false"
        aria-label="Zoom in"
        class="leaflet-control-zoom-in"
        href="#"
        role="button"
        title="Zoom in"
      >
        <span
          aria-hidden="true"
        >
          +
        </span>
      </a>
      <a
        aria-disabled="true"
        aria-label="Zoom out"
        class="leaflet-control-zoom-out leaflet-disabled"
        href="#"
        role="button"
        title="Zoom out"
      >
        <span
          aria-hidden="true"
        >
          −
        </span>
      </a>
    </div>
    <div
      class="leaflet-control test-control"
    >
      <button
        class="test-control-child-1"
      >
        Click Me
      </button>
    </div>
    <div
      class="leaflet-bar leaflet-control"
    >
      <a
        aria-label="Full Screen"
        class="leaflet-control-zoom-fullscreen fullscreen-icon"
        href="#"
        role="button"
        title="Full Screen"
      />
    </div>
  </div>
  <div
    class="leaflet-bottom leaflet-left"
  />
  <div
    class="leaflet-bottom leaflet-right"
  />
</div>
`;

exports[`CustomControls when insertAfterSelector is specified and there is not a node matching that selector, then the control is added at the end of the leaflet-container 1`] = `
<div
  class="leaflet-control-container"
>
  <div
    class="leaflet-top leaflet-left"
  />
  <div
    class="leaflet-top leaflet-right"
  >
    <div
      class="leaflet-control-zoom leaflet-bar leaflet-control"
    >
      <a
        aria-disabled="false"
        aria-label="Zoom in"
        class="leaflet-control-zoom-in"
        href="#"
        role="button"
        title="Zoom in"
      >
        <span
          aria-hidden="true"
        >
          +
        </span>
      </a>
      <a
        aria-disabled="true"
        aria-label="Zoom out"
        class="leaflet-control-zoom-out leaflet-disabled"
        href="#"
        role="button"
        title="Zoom out"
      >
        <span
          aria-hidden="true"
        >
          −
        </span>
      </a>
    </div>
    <div
      class="leaflet-bar leaflet-control"
    >
      <a
        aria-label="Full Screen"
        class="leaflet-control-zoom-fullscreen fullscreen-icon"
        href="#"
        role="button"
        title="Full Screen"
      />
    </div>
    <div
      class="leaflet-control test-control"
    >
      <button
        class="test-control-child-1"
      >
        Click Me
      </button>
    </div>
  </div>
  <div
    class="leaflet-bottom leaflet-left"
  />
  <div
    class="leaflet-bottom leaflet-right"
  />
</div>
`;
